ql/src/java/org/apache/hadoop/hive/ql/parse/MergeSemanticAnalyzer.java

```diff
@@ -188,23 +163,23 @@ WHEN NOT MATCHED THEN INSERT VALUES (source.a2, source.b2)
       switch (getWhenClauseOperation(whenClause).getType()) {
       case HiveParser.TOK_INSERT:
         numInsertClauses++;
-        handleInsert(whenClause, rewrittenQueryStr, targetNameNode, onClause,
+        handleInsert(whenClause, mergeRewriter, targetNameNode, onClause,
```

Comment:
i think we need to extract the whole merge logic. 
check Merge COW patch, I had to pass 3 extra params in every handle method + create extra deleteCow that basically joins deletes from Update and Delete branches

Reply:
I removed `MergeRewriter` interface and implemented `Rewriter` instead and added `MergeStatement` to store the parsed parts of merge.

Reply:
👍